### PR TITLE
feat: passkeys!

### PR DESCRIPTION
**Pull request by hperl**
```
BREAKING CHANGES: This feature enables two-step registration per default. Two-step registration is a significantly improved sign up flow and recommended when using more than one sign up methods. To disable two-step registration, set `selfservice.flows.registration.enable_legacy_flow` to `true`. This value defaults to `false`.
```

## TODOs

- [x] Submit button is on the top of the UI nodes list (should be last item) for API and browser flows
- [x] Social sign in needs to be an option on the first step, not the second one
- [x] Add test case for missing traits with OIDC
- [x] Do not use 422 but instead 400 for API flows and Browser + JSON flows
- [x] Add tests for mobile app
- [x] Back button for sign up with code
- [x] Rename config option to `enable_legacy_one_step`

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
Unfortunately, the PR still suffers from various bugs when both passkey and webauthn strategy are enabled:

**config**

```
log:
  level: trace
  leak_sensitive_values: true
secrets:
  cookie:
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE
selfservice:
  default_browser_return_url: http://localhost:4455/
  allowed_return_urls:
    - http://localhost:4455
    - http://localhost:19006/Callback
    - exp://example.com/Callback
    - https://www.ory.sh/
    - https://example.org/
    - https://www.example.org/
  methods:
    link:
      config:
        lifespan: 1h
    code:
      config:
        lifespan: 1h
    totp:
      enabled: true
      config:
        issuer: issuer.ory.sh
    lookup_secret:
      enabled: true
    webauthn:
      enabled: true
      config:
        passwordless: true
        rp:
          id: localhost
          origins:
            - http://localhost:4455
          display_name: Ory
    passkey:
      enabled: true
      config:
        rp:
          id: localhost
          origins:
            - http://localhost:4455
          display_name: Ory
  flows:
    settings:
      ui_url: http://localhost:4455/settings
      privileged_session_max_age: 5m
      required_aal: aal1
    logout:
      after:
        default_browser_return_url: http://localhost:4455/login
    registration:
      ui_url: http://localhost:4455/registration
      after:
        password:
          hooks:
            - hook: session
        passkey:
          hooks:
            - hook: web_hook
              config:
                body: base64://ZnVuY3Rpb24oY3R4KSBjdHg=
                url: https://webhook-target-gsmwn5ab4a-uc.a.run.app/documents/0.4nwpiet8hx5
                method: PUT
            - hook: session
    login:
      ui_url: http://localhost:4455/login
    error:
      ui_url: http://localhost:4455/error
    verification:
      ui_url: http://localhost:4455/verify
    recovery:
      ui_url: http://localhost:4455/recovery
serve:
  public:
    base_url: http://localhost:4455/
    cors:
      enabled: true
      allowed_origins:
        - http://localhost:3000
        - http://localhost:19006
      allowed_headers:
        - Authorization
        - Content-Type
        - X-Session-Token
  admin:
    base_url: http://kratos:4434/
hashers:
  algorithm: argon2
  argon2:
    memory: 1KB
    iterations: 1
    parallelism: 1
courier:
  smtp:
    connection_uri: smtps://test:test@localhost:1025/?skip_ssl_verify=true
session:
  whoami:
    required_aal: aal1
identity:
  schemas:
    - id: default
      url: file://test/e2e/profiles/passkey/identity.traits.schema.json
```

**traits.json**

```
{
  "$id": "https://schemas.ory.sh/presets/kratos/quickstart/email-password/identity.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Person",
  "type": "object",
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "email": {
          "type": "string",
          "format": "email",
          "title": "Your E-Mail",
          "minLength": 3,
          "ory.sh/kratos": {
            "credentials": {
              "password": {
                "identifier": true
              },
              "webauthn": {
                "identifier": true
              },
              "passkey": {
                "display_name": true
              }
            }
          }
        },
        "website": {
          "title": "Your website",
          "type": "string",
          "format": "uri",
          "minLength": 10
        }
      },
      "required": ["email", "website"],
      "additionalProperties": false
    }
  }
}
```

## Unable to sign in using WebAuthn passwordless strategy

https://github.com/ory/kratos/assets/3372410/d70a24b9-7212-4e68-81e6-129143329e41

Neither hardware keys nor passkeys work when I click "Use security key".

## Settings page shows duplicate entries

"Add security key" is shown twice, and the hardware key is missing from the list:

Screenshot 2024-02-12 at 12 46 25

## Hardware key is hidden from view when it is the only method

Only when I add two hardware keys, the keys are actually shown:

Screenshot 2024-02-12 at 12 49 48

Instead, I would expect to see the key with a disabled remove button.

## Left to be done

Can you please add e2e test cases for these and other cases? In my view these bugs are there because the following test suites are missing:

- Migration scenarios where both PassKey and WebAuthn + Passwordless are enabled
- Using the keys interchangeably (use WebAuthn keys to sign in using PassKey strategy) during login
- Changing settings while both strategies are enabled (add, delete)

## Documentation and migration guide

Side note: This is what it looks like when you sign in using WebAuthn + Passwordless with a hardware key, disable WebAuthn and just have Passkey enabled, and then try to sign in using the hardware key:

Screenshot 2024-02-12 at 12 53 19

I think we should document this limitation somewhere.

**Comment by aeneasr** [comment]
For two-step registration, there seems to be an issue:

Screenshot 2024-02-12 at 12 59 18
Screenshot 2024-02-12 at 12 59 29

For two-step registration I would also appreciate to see a couple more use cases where several different sign up methods are used to ensure that all of them work properly (including scenarios where the form doesn't validate).

File: persistence/sql/identity/persister_identity.go [hunk 1]
```diff
@@ -228,6 +229,36 @@
 	return i.CopyWithoutCredentials(), creds, nil
 }
 
+func (p *IdentityPersister) FindIdentityByWebauthnUserHandle(ctx context.Context, userHandle string) (_ *identity.Identity, err error) {
```
[the remark this answers is not in the document]
Comment (hperl) [reply persistence/sql/identity/persister_identity.go:245]: done

**Comment by aeneasr** [comment]
To do:

1. Submit button is on the top of the UI nodes list (should be last item) for API and browser flows
2. Social sign in needs to be an option on the first step, not the second one:
   * Add test case for missing traits
3. Do not use 422 but instead 400 for API flows and Browser + JSON flows
4. Add tests for mobile app
5. Back button for sign up with code

**Comment by aeneasr** [comment]
I unfortunately still receive an error when both passkeys + webauthn with passwordless are enabled and I try to sign up using webauthn:

Screenshot 2024-02-22 at 13 44 40

**Review by aeneasr — APPROVED** [review]
Amazing job

**Comment by bcspragu** [comment]
Hi there,

First time Ory-er (Kratos-er?) here, thanks for building great software! I think some of the docs need to be updated ([example](https://www.ory.sh/docs/kratos/bring-your-own-ui/custom-ui-basic-integration#create-registration-flow)), because they show the `createRegistrationFlow` response containing a password field, and I could not figure out for the life of me why I wasn't getting a password field in my response with the default identity schema.

After enabling `selfservice.flows.registration.enable_legacy_one_step`, the password field showed up in the response.

**Pull request merged**